test/e2e/knativeservice_test.go

```diff
@@ -0,0 +1,120 @@
+package e2e
```

Comment:
@pratikjagrut Considering the tests executed against Knative resources, we should move those as unit-tests. Since there's no backing operator for Knative services, we would not see the benefits of writing an end-to-end test for this purpose, and in unit-tests level we are completely able to validate contracts, in the same line than @DhritiShikhar's PR #247.

Reply:
Perhaps a little bit more clarity is required here.

The reason this test makes sense to be considered a unit test is that it in its current form doesn't exercise much more than those, but with the burden of having to provision a cluster. 

Using the existing libraries made available by the operator-sdk to marshal/unmarshal manifests are the same used in the API server, so exercising marshalling and unmarshalling is enough to guarantee compatibility between different systems; in this particular case would be modifying Knative manifest to enable the binding in such a way it is properly encoded (the value encoded fits a given type) and semantically valid (against OpenAPI validation rules declared in the CRD). 

Producing properly encoded and valid changes in the API server is what we are pursuing at each reconciliation, and for this we need to rely on the assumption the other system will work. I find end-to-end tests really valuable to confirm that all integrated systems behave as expected; please note this is fundamentally different from unit tests, which is to prove the contract is correct.

Reply:
@isutton so this test can be directly moved under unit test as it is?

Reply:
@pratikjagrut I'm not sure about the "as is" part. If you have access from `Reconcile()` to the data you expect to exist then executing `Reconcile()` directly with a client that already has all the manifests you expect to exist in the cluster is as effective as if you're testing against a cluster (assuming the contract is being properly tested, it is).

Given that, configuring a client with supporting manifests (all the manifests that are expected to be consulted during reconciliation) to be tested and exercising the business logic by triggering the reconciliation process is an approach I sympathize with.

Reply:
I pushed the unit test code. Resolving this conversation.